File: app/models/tag.rb
class Tag

  class NotFound < StandardError; end

  attr_reader :name, :pages

  def self.all
    mapping = Hash.new { |hash, key| hash[key] = [] }
    Page.all.each do |page|
      page.tags.each { |tag| mapping[tag] << page }
    end

    tags = []
    mapping.each do |name, pages|
      tags << Tag.new(name, pages)
    end
    tags.sort { |a,b| a.name <=> b.name }
  end

  def self.find(name)
    tag = all.find { |tag| tag.name == name }
    raise NotFound, "unable to find tag with name #{name.inspect}" if tag.nil?
    tag
  end

  def initialize(name, pages)
    @name = name
    @pages = pages
  end

  def to_param
    name
  end

  def page_count
    pages.count
  end

  def sorted_pages
    pages.sort { |a,b| a.name <=> b.name }
  end

  def last_modified
    pages.map(&:mtime).sort.last
  end

end

File: src/tag.rs
use std::collections::BTreeMap;
use std::path::Path;
use std::rc::Rc;
use std::time::SystemTime;

use crate::page::{Loaded, Page};

pub type Pages = Vec<Rc<Page<Loaded>>>;

#[derive(Debug)]
pub struct Tag {
    pub name: String,
    pages: Pages,
}

impl Tag {
    fn new(name: String, pages: Pages) -> Tag {
        Tag { name, pages }
    }

    pub(crate) fn all(basepath: &Path) -> Vec<Tag> {
        let mappings = Page::all(basepath)
            .into_iter()
            .filter_map(|page| page.load().ok())
            .filter_map(|page| {
                if page.is_hidden() {
                    None
                } else {
                    Some(Rc::new(page))
                }
            })
            .fold(
                BTreeMap::new(),
                |mut mappings: BTreeMap<String, Pages>, page| {
                    if !page.is_hidden() {
                        page.tags().iter().for_each(|name| {
                            mappings
                                .entry(name.to_string())
                                .or_default()
                                .push(Rc::clone(&page))
                        });
                        mappings
                    } else {
                        mappings
                    }
                },
            );

        mappings
            .into_iter()
            .map(|(name, pages)| Tag::new(name, pages))
            .collect()
    }

    pub fn find(name: &str, basepath: &Path) -> Option<Tag> {
        Tag::all(basepath).into_iter().find(|tag| tag.name == name)
    }

    pub fn page_count(&self) -> usize {
        self.pages.len()
    }

    pub fn sorted_pages(&self) -> Pages {
        let mut sorted = self.pages.clone();
        sorted.sort_by(|a, b| a.name.cmp(&b.name));
        sorted
    }

    pub fn last_modified(&self) -> SystemTime {
        self.pages
            .iter()
            .map(|page| page.mtime())
            .max()
            .unwrap_or_else(|| SystemTime::now())
    }
}
